eth/catalyst/api.go

```diff
@@ -105,9 +173,105 @@ func (api *consensusAPI) makeEnv(parent *types.Block, header *types.Header) (*bl
 	return env, nil
 }
 
+func (api *ConsensusAPI) PreparePayload(params AssembleBlockParams) (*PayloadResponse, error) {
+	data, err := api.assembleBlock(params)
+	if err != nil {
+		return nil, err
+	}
+	id := len(api.preparedBlocks)
+	api.preparedBlocks[id] = data
+	return &PayloadResponse{PayloadID: uint64(id)}, nil
+}
+
+func (api *ConsensusAPI) GetPayload(PayloadID hexutil.Uint64) (*ExecutableData, error) {
+	data, ok := api.preparedBlocks[int(PayloadID)]
+	if !ok {
+		return nil, &UnknownPayload
+	}
+	return data, nil
+}
+
+// ConsensusValidated is called to mark a block as valid, so
+// that data that is no longer needed can be removed.
+func (api *ConsensusAPI) ConsensusValidated(params ConsensusValidatedParams) error {
+	switch params.Status {
+	case VALID.Status:
+		// Finalize the transition if it's the first `FinalisedBlock` event.
+		merger := api.merger()
+		if !merger.EnteredPoS() {
+			if err := api.checkTerminalTotalDifficulty(params.BlockHash); err != nil {
+				return err
+			}
+			merger.EnterPoS()
+		}
+		return nil //api.setHead(params.BlockHash)
+	case INVALID.Status:
+		// TODO (MariusVanDerWijden) delete the block from the bc
+		return nil
+	default:
+		return errors.New("invalid params.status")
+	}
+}
+
+func (api *ConsensusAPI) ForkchoiceUpdated(params ForkChoiceParams) error {
+	var emptyHash = common.Hash{}
+	if !bytes.Equal(params.HeadBlockHash[:], emptyHash[:]) {
+		if err := api.checkTerminalTotalDifficulty(params.HeadBlockHash); err != nil {
+			return err
+		}
+		return api.setHead(params.HeadBlockHash)
+	}
+	return nil
+}
+
+// ExecutePayload creates an Eth1 block, inserts it in the chain, and returns the status of the chain.
+func (api *ConsensusAPI) ExecutePayload(params ExecutableData) (GenericStringResponse, error) {
+	if api.light {
+		parent := api.les.BlockChain().GetHeaderByHash(params.ParentHash)
+		if parent == nil {
+			return INVALID, fmt.Errorf("could not find parent %x", params.ParentHash)
+		}
+		block, err := ExecutableDataToBlock(api.les.BlockChain().Config(), parent, params)
+		if err != nil {
+			return INVALID, err
+		}
+		if err = api.les.BlockChain().InsertHeader(block.Header()); err != nil {
+			return INVALID, err
+		}
+		return VALID, nil
+	}
+	parent := api.eth.BlockChain().GetBlockByHash(params.ParentHash)
+	if parent == nil {
+		return INVALID, fmt.Errorf("could not find parent %x", params.ParentHash)
```

Comment:
I don't think this will work. Our RPC server implementation, AFAIK, when dealing with a non-nil return and a non-nil error, will discard the return vaue and return only the error. This might even be a 'feature' of json rpc: either you return a response XOR an error. Our server chooses the error.